nixos/modules/config/nix-channel.nix

```diff
@@ -94,10 +95,22 @@ in
       NIX_PATH = cfg.nixPath;
     };
 
-    nix.settings.nix-path = mkIf (! cfg.channel.enable) (mkDefault "");
-
     systemd.tmpfiles.rules = lib.mkIf cfg.channel.enable [
       ''f /root/.nix-channels - - - - ${config.system.defaultChannel} nixos\n''
     ];
+
+    system.activationScripts.no-nix-channel = mkIf (!cfg.channel.enable)
+      (stringAfter [ "etc" "users" ] ''
+        if [ -e "/root/.nix-defexpr/channels" ]; then
+            echo "WARNING: /root/.nix-defexpr/channels exists, but channels have been disabled." 1>&2
+            echo "Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset." 1>&2
+            echo "Delete the above directory to prevent this." 1>&2
+        fi
+        if [ -e "/nix/var/nix/profiles/per-user/root/channels" ]; then
+            echo "WARNING: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled." 1>&2
+            echo "Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset." 1>&2
+            echo "Delete the above directory to prevent this." 1>&2
+        fi
+      '');
```

Comment:
Shouldn't we also check for user channels?

Reply:
Yes, good idea. Even though we can't enumerate all possible values of `HOME` that Nix could be invoked with, we could at least check the home directories in the user catalog. Added.

Reply:
That easily catches 99% of users in my estimate, and that's a big win. :+1: